### DISCUSSION TITLE
Cannot invoke DELETE, get a 404

### DISCUSSION THREAD

**Discussion by shaktech786** [vercel/next.js#57206] [Help] [answered]
### Summary

I'm using the app router system (not pages) and am able to execute GET and POSTs from route.js using the fetch api successfully. However, when I try to invoke a DELETE, I get a 404 with no additional information nor any useful logs in my terminal. I believe my fetch and DELETE functions are written correctly (see below). My folder structure looks correct as well (/src/api/artists/route.js). Not sure what else to do and I couldn't find the same issue online anywhere. Would appreciate any help, thank you!

### Additional information

```js
// fetch call when I click a button to delete
const handleDeleteClick = async (artist) => {
    try {
      const response = await fetch(`/api/artists/${artist.artist_id}`, {
        method: "DELETE"
      });

      if (response.ok) {
        setSelectedArtist(null);
        setModalIsOpen(false);
      } else {
        console.log(response);
        // this console log prints
        console.error("Failed to delete artist");
      }
    } catch (error) {
      console.error(error);
    }
  };

// from route.js:
export async function DELETE(req) {
  console.log("this print never gets called")
  try {
    const artistId = req.query.id;
    // await Artist.destroy({ where: { id: artistId } });
    return NextResponse.redirect(new URL("/", req.url));
  } catch (error) {
    console.error("Unable to delete artist:", error);
    return NextResponse.error(error);
  }
}

// terminal output:
 ○ Compiling /not-found ...
 ✓ Compiled /not-found in 4.2s (421 modules)
```

**Comment by reikal1337** [comment]
Try to print artist and see if it always contains "artist_id".
Try adding localhost to fetch. Like this `http://localhost:3000/api/artists/${artist.artist_id}`.
is your route.js file inside [id] folder ?
Also can you show how you trigger handleDeleteClick() ?

**Reply by shaktech786** [reply]
Thanks for your response!

Artist_id does always print correctly, adding localhost didn't help, and also wasn't needed for the GET/POST. What should the [id] folder structure look like? I couldn't find any docs on it. Currently my path is /src/app/api/artists/route.js. Route.js contains my GET, POST, and DELETE.

handleDeleteClick is triggered here:

```
<tbody>
          {filteredArtists.map((artist) => (
            <tr
              key={artist.artist_id}
              onClick={() => handleRowClick(artist)}
              className="cursor-pointer transition-colors hover:bg-gray-100"
            >
              <td className="py-2 px-4">{artist.first_name}</td>
              <td className="py-2 px-4">{artist.last_name}</td>
              <td className="py-2 px-4">{artist.art_type}</td>
              <td className="py-2 px-4">{artist.jamat_khana}</td>
              <td className="py-2 px-4">{artist.city}</td>
              <td className="py-2 px-4">{artist.region}</td>
              <td className="py-2 px-4">{artist.country}</td>
              <td className="py-2 px-4">
                <button
                  onClick={(event) => {
                    event.stopPropagation();
                    handleDeleteClick(artist);
                  }}
                  className="text-red-500 hover:text-red-700"
                >
                  X
                </button>
              </td>
            </tr>
          ))}
        </tbody>
```

**Comment by reikal1337** [comment, marked as answer]
Not exactly sure where the problem lies, but you should use [Dynamic Route Segments](https://nextjs.org/docs/app/building-your-application/routing/route-handlers#dynamic-route-segments) and have id as param instead of query.
 Folder structure would look like "/src/app/api/artists/[id]/route.js." [id] is a folder.
Move everything that passes or needs a id in the route inside "[id]" folder. So GET,DELETE or PATCH for specific artist here. Then your del should look like:

```
export async function DELETE(req, { params }) {
  console.log("this print never gets called")
  try {
    const { id } = params
    // await Artist.destroy({ where: { id: id} });
    return NextResponse.redirect(new URL("/", req.url));
  } catch (error) {
    console.error("Unable to delete artist:", error);
    return NextResponse.error(error);
  }
}
```

Hope it helps.

**Reply by shaktech786** [reply]
Didn't understand the [id] folder thing before, got it now thanks!
